### PR TITLE
Feat/ade 499/idea tech dao create update

### PR DESCRIPTION
**Pull request by robinMcA**

### CODE REVIEW

File: src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+   * Create new entry to represent the a FedexIdea
+   * @param fedexIdea FedexIdea model object
+   * @return FedexIdea object created in data store
+   */
+  public FedexIdea create(FedexIdea fedexIdea) {
+    AoFedexIdea aoFedexIdea =
+      this.ao.create(AO_FEDEX_IDEA_TYPE);
+    List<AoFedexTechnology> aoTechList = new ArrayList<>();
+
+    fedexIdea
```
Comment (dspasojevic) [inline src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java:51]: Why is this broken up before and after the save? Maybe a comment would explain.

File: src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+  /**
+   * Convert array of active objects to a list of model objects
+   * @param aoFedexIdeas AoFedexIdea[]
+   * @return List<FedexIdea>
+   */
+  private List<FedexIdea> asListFedexIdea(AoFedexIdea[] aoFedexIdeas) {
+    List<FedexIdea> ideas = Lists.newArrayList();
+
+    for (AoFedexIdea aoFedexIdea : aoFedexIdeas) {
```
Comment (dspasojevic) [inline src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java:149]: `map`?

File: src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+   * Convert array of active objects to a list of model objects
+   * @param aoFedexTechnologies AoFedexTechnology[]
+   * @return List<FedexTechnology>
+   */
+  private List<FedexTechnology> asListFedexTechnology(
+    AoFedexTechnology[] aoFedexTechnologies
+  ) {
+    List<FedexTechnology> tech = Lists.newArrayList();
+
+    for (AoFedexTechnology aoFedexTechnology : aoFedexTechnologies) {
```
Comment (dspasojevic) [inline src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java:166]: `map`?

File: src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+   * @return FedexTechnology object
+   */
+  private FedexTechnology asFedexTechnology(
+    AoFedexTechnology aoFedexTechnology
+  ) {
+    return aoFedexTechnology == null
+      ? null
+      : (
+        new au.com.agiledigital.idea_search.model.FedexTechnology.Builder()
+      ).withGlobalId(aoFedexTechnology.getGlobalId())
```
Comment (dspasojevic) [inline src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java:261]: Are these parens necessary?

File: src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    //        If the title of the page already exists the create event fails and a page update event is fired on a successful save,
+    //        if this happens the aoFedexIdea will not exist in the data store and will need to be created.
+    if (aoFedexIdea == null) {
+      aoFedexIdea = this.ao.create(AO_FEDEX_IDEA_TYPE);
+    }
+
+    List<AoFedexTechnology> aoFedexTechnology = Arrays.asList(
+      aoFedexIdea.getTechnology()
+    );
+    aoFedexTechnology.forEach(this.ao::delete);
```
Comment (dspasojevic) [inline src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java:93]: This just deletes the association between the idea and the tech?

File: src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+    this.prepareAOFedexIdea(aoFedexIdea, fedexIdea);
+
+    aoFedexIdea.save();
+
+    final AoFedexIdea idea = aoFedexIdea;
+
+    aoTechList.forEach(
+      li -> {
+        li.setIdea(idea);
```
Comment (dspasojevic) [inline src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java:112]: This looks like you are managing the join table manually. AO doesn't help you with this?

File: src/main/java/au/com/agiledigital/idea_search/listener/FedexIdeaEventListener.java [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.w3c.dom.ls.DOMImplementationLS;
+import org.w3c.dom.ls.LSSerializer;
+import org.xml.sax.SAXException;
+
+@Named
+public class FedexIdeaEventListener
```
Comment (dspasojevic) [inline src/main/java/au/com/agiledigital/idea_search/listener/FedexIdeaEventListener.java:41]: Comment.

File: src/main/java/au/com/agiledigital/idea_search/listener/FedexIdeaEventListener.java [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+    String moduleCompleteKey = event.getBlueprint().getModuleCompleteKey();
+
+    String thing = MY_BLUEPRINT_KEY.getCompleteKey();
+
+    if (thing.equals(moduleCompleteKey)) {
+      try {
+        FedexIdea idea = getFedexIdea(event.getPage());
+        this.fedexIdeaService.create(idea);
+      } catch (ParserConfigurationException | IOException | SAXException e) {
+        e.printStackTrace();
```
Comment (dspasojevic) [inline src/main/java/au/com/agiledigital/idea_search/listener/FedexIdeaEventListener.java:153]: Is there a better logging option?

File: src/main/resources/META-INF/spring/spring-components.xml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        <beans:constructor-arg name="fedexIdeaDao" ref="fedexIdeaDao" />
+    </beans:bean>
+    <osgi:service id="fedexIdeaService_osgiService" ref="fedexIdeaService">
+        <osgi:interfaces>
+            <beans:value
+      >au.com.agiledigital.idea_search.service.FedexIdeaService</beans:value>
+        </osgi:interfaces>
+    </osgi:service>
+
+<!--    <beans:bean id="FedexIdeaDao" autowire="default" class="au.com.agiledigital.idea_search.dao.FedexIdeaDao">-->
```
Comment (dspasojevic) [inline src/main/resources/META-INF/spring/spring-components.xml:24]: ?

File: src/main/java/au/com/agiledigital/idea_search/listener/FedexIdeaEventListener.java (hunk 8 above)
Comment (robinMcA) [reply src/main/java/au/com/agiledigital/idea_search/listener/FedexIdeaEventListener.java]: using slf4j now with debugs errors

File: src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java (hunk 4 above)
Comment (robinMcA) [reply src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java]: No

File: src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+        return fedexIdea
+                .getTechnologies()
+                .stream()
+                .map(tech -> this.prepareAOFedexTechnology(tech.getTechnology()))
+                .collect(Collectors.toList());
+    }
+
+    //    See comment on https://community.atlassian.com/t5/Jira-questions/ActiveObjects-jira/qaq-p/354375
+    //    This means that the setter is done on the Recipient entity(setting the Filter) and because of the way relational databases...
+    private void SetTechnology(
```
Comment (SamMurphyDev) [inline src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java:121]: Lowercase S or is this forced by Confluence?

```suggestion
    private void setTechnology(
```

File: src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+    public List<FedexTechnology> findAllTech() {
+        AoFedexTechnology[] aoFedexTechnologies =
+                this.ao.find(AO_FEDEX_TECHNOLOGY_TYPE, Query.select());
+        return this.asListFedexTechnology(aoFedexTechnologies);
+    }
+
+    /**
+     * Convert array of active objects to a list of model objects
+     *
+     * @param aoFedexIdeas AoFedexIdea[]
```
Comment (SamMurphyDev) [inline src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java:157]: Should this be
```suggestion
     * @param aoFedexIdeas Some description here?
```
As the type information is inferred?

File: src/main/java/au/com/agiledigital/idea_search/helpers/MacroHelpers.java [hunk 12]
```diff
@@ -10,12 +10,15 @@
  * Helpers to do data transformation for macros
  */
 public class MacroHelpers {
+
   public static Set<String> splitTrimToSet(String str, String delimiter) {
```
Comment (SamMurphyDev) [inline src/main/java/au/com/agiledigital/idea_search/helpers/MacroHelpers.java:14]: I know this was me who wrote this, but document public function

```suggestion
  /**
   * Splits a string at a delimiter, then remove the starting and ending whitespace from each piece
   * @param str value to be split
   * @param delimiter search key to split the string on
   * @return Set of the sections
   */
  public static Set<String> splitTrimToSet(String str, String delimiter) {
```

File: src/main/resources/atlassian-plugin.xml [hunk 13]
```diff
@@ -35,6 +35,7 @@
         template-key="ideaSearch.blueprint.page.fedexIdea.page1Form"
         title-key="ideaSearch.blueprint.page.wizard.page1.title"
       />
+      />
```
Comment (SamMurphyDev) [inline src/main/resources/atlassian-plugin.xml:38]: Is this meant to be here?

**Review by SamMurphyDev — COMMENTED** [review]
Just a couple of nitpicks

File: src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java (hunk 10 above)
Comment (robinMcA) [reply src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java]: setTechnologies

File: src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java (hunk 11 above)
Comment (robinMcA) [reply src/main/java/au/com/agiledigital/idea_search/dao/FedexIdeaDao.java]: added

File: src/main/java/au/com/agiledigital/idea_search/helpers/MacroHelpers.java (hunk 12 above)
Comment (robinMcA) [reply src/main/java/au/com/agiledigital/idea_search/helpers/MacroHelpers.java:20]: added

File: src/main/resources/atlassian-plugin.xml (hunk 13 above)
Comment (robinMcA) [reply src/main/resources/atlassian-plugin.xml]: :eyes: no. deleted

**Review by SamMurphyDev — APPROVED** [review]
LGTM

**Pull request merged**